### PR TITLE
Revise documentation of Minibatch

### PR DESCRIPTION
**Pull request by rpgoldman**
I was trying to educate myself about Minibatch, and when I looked at the page in the documentation, I found that it was formatted wrong.
I revised it to use Numpy doc style, and added the `sphinx.ext.napoleon` Sphinx extension to properly format numpy-style docstrings.
There were some things I didn't understand (my understanding of theano is limited, but that's probably ok, since most users will start out that way), which I have marked with FIXME. If someone will clue me in about these, I will clear them up.

### CODE REVIEW

File: pymc3/data.py [hunk 1]
```diff
@@ -127,109 +128,146 @@ class Minibatch(tt.TensorVariable):
...
-    3) take that slice
-    >>> minibatch = shared[ridx]
+        >>> minibatch = shared[ridx]
 
-    That's done. Next you can use this minibatch somewhere else.
-    You can see that implementation does not require fixed shape
-    for shared variable. Feel free to use that if needed.
+    That's done. Now you can use this minibatch somewhere else.
+    You can see that the implementation does not require a fixed shape
+    for the shared variable. Feel free to use that if needed.
```
Comment (ferrine) [inline pymc3/data.py:235]: "feel free to use the created minibatch where it fits your needs"

File: pymc3/data.py [hunk 2]
```diff
@@ -127,109 +128,146 @@ class Minibatch(tt.TensorVariable):
...
+    Then you should create a `dict` with replacements:
+
     >>> replacements = {x: testdata}
     >>> rnode = theano.clone(node, replacements)
     >>> assert (testdata ** 2 == rnode.eval()).all()
 
-    To replace minibatch with it's shared variable you should do
-    the same things. Minibatch variable is accessible as an attribute
-    as well as shared, associated with minibatch
+    *FIXME: In the following, what is the **reason** to replace the Minibatch variable with
```
Comment (ferrine) [inline pymc3/data.py:250]: The reason is to make computations non random for the full dataset

File: pymc3/data.py (hunk 1 above)
Comment (ferrine) [reply pymc3/data.py:235]: Probably worth noting that Minibatch should not go along with most MCMC algos and mostly for VI

File: pymc3/data.py [hunk 3]
```diff
@@ -127,109 +128,146 @@ class Minibatch(tt.TensorVariable):
 
     Notes
     -----
-    Below is a common use case of Minibatch within the variational inference.
-    Importantly, we need to make PyMC3 "aware" of minibatch being used in inference.
+    Below is a common use case of Minibatch with variational inference.
```
Comment (ferrine) [inline pymc3/data.py:146]: Minibatching is a common usecase in variational inference

**Review by abhinavsagar — APPROVED** [review]
LGTM

**Comment by twiecki** [comment]
@rpgoldman should we try and merge this?

**Comment by rpgoldman** [comment]
@twiecki I believe the reason I called this just a WIP is that all I was able to do was to revise the existing documentation.  I really didn't know what Minibatch was *for*, and still don't, so my edits are obviously sub-optimal.  I gather it's something to do with VI, which I haven't used myself (my use of PyMC3 and my edits to the code base are shaped by what I am doing with it).

**Comment by rpgoldman** [comment]
@twiecki I tweaked this to make the merge work.  I still note that this bit of the documentation is ... impossible for me to parse:

    Important note: :class:``Minibatch`` has ``shared``, and ``minibatch`` attributes
    you can call later:

    >>> x.set_value(np.random.laplace(size=(100, 100)))

    and minibatches will be then from new storage
    it directly affects ``x.shared``.

This is an example of something I didn't know well enough to edit.  Maybe someone else?

**Comment by ferrine** [comment]
This is an example of how to manipulate minibatch underlying storage.

**Comment by rpgoldman** [comment]
> This is an example of how to manipulate minibatch underlying storage.

The problem with this as documentation is that either you know that `Minibatch` has some kind of special underlying storage, and then this doesn't help, or you don't know, in case this doesn't help.
Similarly, the one line summary is

    Multidimensional minibatch that is pure TensorVariable

which really isn't helpful because

1. it defines minibatch in terms of minibatch

2. it assumes that the reader knows what a `TensorVariable` is, and what it means for something to be a pure `TensorVariable` instead of an impure (?) one.

3. It says "Multidimensional minibatch," leaving the reader wondering what a one-dimensional minibatch would be.

4. Does this mean that the minibatch is a *single* `TensorVariable`, or that it is *made up of* `TensorVariable`s? (From the source it is clear that it is the former, but see number 2 above).


What this really needs is a one line definition that states what a minibatch is in a non-circular way.

**Comment by rpgoldman** [comment]
@ferrine and I both made concurrent changes.  I have merged them in this most recent push, which also rebases.

@twiecki if this passes, how about merging it?

I think the docstrings could still be better, but these are an improvement over what's there, so let's get it incorporated. Thanks!

**Comment by rpgoldman** [comment]
@ferrine Can we merge this?

**Comment by twiecki** [comment]
Thanks!

**Pull request merged**